resource/aws_ec2_managed_prefix_list: The plugin encountered an error, and failed to respond to the

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform CLI and Terraform AWS Provider Version

$ terraform -v
Terraform v0.15.3
on darwin_amd64
+ provider registry.terraform.io/hashicorp/aws v3.40.0

### Affected Resource(s)

* aws_ec2_managed_prefix_list

### Terraform Configuration Files

None.

### Debug Output

None.

### Panic Output

https://gist.github.com/tcharewicz/309944f89409bcff10cbeada57db0331

### Expected Behavior

Terraform should modify resouce, add or delete description.

```hcl
  # aws_ec2_managed_prefix_list.this will be updated in-place
  ~ resource "aws_ec2_managed_prefix_list" "this" {
        id             = "pl-0f8022f4f29018d7a"
        name           = "this"
        tags           = {
            "ResourceOwner" = "test"
        }
      ~ version        = 1 -> (known after apply)
        # (5 unchanged attributes hidden)

      + entry {
          + cidr = "1.1.1.0/24"
        }
      - entry {
          - cidr        = "1.1.1.0/24" -> null
          - description = "this" -> null
        }
      + entry {
          + cidr = "1.1.2.0/24"
        }
      - entry {
          - cidr        = "1.1.2.0/24" -> null
          - description = "this" -> null
        }
      + entry {
          + cidr = "1.1.3.0/24"
        }
      - entry {
          - cidr        = "1.1.3.0/24" -> null
          - description = "this" -> null
        }
      + entry {
          + cidr = "1.1.4.0/24"
        }
      - entry {
          - cidr        = "1.1.4.0/24" -> null
          - description = "this" -> null
        }
      + entry {
          + cidr = "1.1.5.0/24"
        }
      - entry {
          - cidr        = "1.1.5.0/24" -> null
          - description = "this" -> null
        }
    }

Plan: 0 to add, 1 to change, 0 to destroy.
```

### Actual Behavior

Terraform crashed.

### Steps to Reproduce

1. Create manged prefix list, with 5 entries. All entries has descriptions.

```hcl
provider "aws" {
  region = "eu-central-1"
}

resource "aws_ec2_managed_prefix_list" "this" {
  name           = "this"
  address_family = "IPv4"
  max_entries    = 5

  entry {
    cidr = "1.1.1.0/24"
    description = "this"
  }

  entry {
    cidr = "1.1.2.0/24"
    description = "this"
  }

  entry {
    cidr = "1.1.3.0/24"
    description = "this"
  }

  entry {
    cidr = "1.1.4.0/24"
    description = "this"
  }

  entry {
    cidr = "1.1.5.0/24"
    description = "this"
  }

  tags = {
      "ResourceOwner"      = "test"
  }
}
```

2. Applay configuration

```
terraform init
terraform apply
```

3. Comment descriptions at entries level.

```hcl
provider "aws" {
  region = "eu-central-1"
}

resource "aws_ec2_managed_prefix_list" "this" {
  name           = "this"
  address_family = "IPv4"
  max_entries    = 5

  entry {
    cidr = "1.1.1.0/24"
    # description = "this"
  }

  entry {
    cidr = "1.1.2.0/24"
    # description = "this"
  }

  entry {
    cidr = "1.1.3.0/24"
    # description = "this"
  }

  entry {
    cidr = "1.1.4.0/24"
    # description = "this"
  }

  entry {
    cidr = "1.1.5.0/24"
    # description = "this"
  }

  tags = {
      "ResourceOwner"      = "test"
  }
}
```

4. Apply

```
terraform apply
```

### Important Factoids

None.

### References

Three weeks ago there were a PR, that changed Behavior of that resource.

https://github.com/hashicorp/terraform-provider-aws/pull/19095

The same behavior you can observe at TF version 0.12.31 and 0.14.11. I do not make a test with TF 0.13.

I can reproduce this with a slightly modified acceptance test:

```console
% make testacc TEST=./aws TESTARGS='-run=TestAccAwsEc2ManagedPrefixList_Entry_Description'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAwsEc2ManagedPrefixList_Entry_Description -timeout 180m
=== RUN   TestAccAwsEc2ManagedPrefixList_Entry_Description
=== PAUSE TestAccAwsEc2ManagedPrefixList_Entry_Description
=== CONT  TestAccAwsEc2ManagedPrefixList_Entry_Description
panic: runtime error: index out of range [1] with length 1

goroutine 1081 [running]:
github.com/terraform-providers/terraform-provider-aws/aws.resourceAwsEc2ManagedPrefixListUpdate(0xc0021f6680, 0x8047f40, 0xc002b52000, 0x0, 0xffffffffffffffff)
	/Users/ewbankkit/src/github.com/terraform-providers/terraform-provider-aws/aws/resource_aws_ec2_managed_prefix_list.go:253 +0x1e4a
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).update(0xc0016a4620, 0x9f182a8, 0xc0020e65c0, 0xc0021f6680, 0x8047f40, 0xc002b52000, 0x0, 0x0, 0x0)
	/Users/ewbankkit/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.6.1/helper/schema/resource.go:352 +0x1ee
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0xc0016a4620, 0x9f182a8, 0xc0020e65c0, 0xc002292690, 0xc00416a9c0, 0x8047f40, 0xc002b52000, 0xc003162130, 0x7e5a8a0, 0xc0022c0c00, ...)
	/Users/ewbankkit/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.6.1/helper/schema/resource.go:464 +0x390
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0xc003c26b70, 0x9f182a8, 0xc0020e65c0, 0xc00239f900, 0xc0020e65c0, 0x8c51dc0, 0xc0022ade00)
	/Users/ewbankkit/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.6.1/helper/schema/grpc_provider.go:955 +0x8ef
github.com/hashicorp/terraform-plugin-go/tfprotov5/server.(*server).ApplyResourceChange(0xc002760260, 0x9f18350, 0xc0020e65c0, 0xc0022925b0, 0xc002760260, 0xc0022ade90, 0xc003cccba0)
	/Users/ewbankkit/go/pkg/mod/github.com/hashicorp/terraform-plugin-go@v0.3.0/tfprotov5/server/server.go:332 +0xb5
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ApplyResourceChange_Handler(0x8c51dc0, 0xc002760260, 0x9f18350, 0xc0022ade90, 0xc002ee8480, 0x0, 0x9f18350, 0xc0022ade90, 0xc003ca8800, 0x381)
	/Users/ewbankkit/go/pkg/mod/github.com/hashicorp/terraform-plugin-go@v0.3.0/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:380 +0x214
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0038eafc0, 0x9f42498, 0xc0048c3680, 0xc0021f2500, 0xc001d92540, 0xdca9c60, 0x0, 0x0, 0x0)
	/Users/ewbankkit/go/pkg/mod/google.golang.org/grpc@v1.32.0/server.go:1194 +0x52b
google.golang.org/grpc.(*Server).handleStream(0xc0038eafc0, 0x9f42498, 0xc0048c3680, 0xc0021f2500, 0x0)
	/Users/ewbankkit/go/pkg/mod/google.golang.org/grpc@v1.32.0/server.go:1517 +0xd0c
google.golang.org/grpc.(*Server).serveStreams.func1.2(0xc000d00600, 0xc0038eafc0, 0x9f42498, 0xc0048c3680, 0xc0021f2500)
	/Users/ewbankkit/go/pkg/mod/google.golang.org/grpc@v1.32.0/server.go:859 +0xab
created by google.golang.org/grpc.(*Server).serveStreams.func1
	/Users/ewbankkit/go/pkg/mod/google.golang.org/grpc@v1.32.0/server.go:857 +0x1fd
FAIL	github.com/terraform-providers/terraform-provider-aws/aws	20.161s
FAIL
make: *** [testacc] Error 1
```

--- a/.changelog/19517.txt
+++ b/.changelog/19517.txt
@@ -1,0 +1,3 @@
+```release-note:bug
+resource/aws_ec2_managed_prefix_list: Fix crash with multiple description-only updates
+```

--- a/aws/resource_aws_ec2_managed_prefix_list.go
+++ b/aws/resource_aws_ec2_managed_prefix_list.go
@@ -242,22 +242,32 @@ func resourceAwsEc2ManagedPrefixListUpdate(d *schema.ResourceData, meta interfac
 		// one with a collection of all description-only removals and the
 		// second one will add them all back.
 		if len(input.AddEntries) > 0 && len(input.RemoveEntries) > 0 {
-			removalInput := &ec2.ModifyManagedPrefixListInput{
-				CurrentVersion: input.CurrentVersion,
-				PrefixListId:   aws.String(d.Id()),
-			}
+			descriptionOnlyRemovals := []*ec2.RemovePrefixListEntry{}
+			removals := []*ec2.RemovePrefixListEntry{}
 
-			for idx, removeEntry := range input.RemoveEntries {
+			for _, removeEntry := range input.RemoveEntries {
+				inAddAndRemove := false
+
 				for _, addEntry := range input.AddEntries {
 					if aws.StringValue(addEntry.Cidr) == aws.StringValue(removeEntry.Cidr) {
-						removalInput.RemoveEntries = append(removalInput.RemoveEntries, input.RemoveEntries[idx])
-						input.RemoveEntries = append(input.RemoveEntries[:idx], input.RemoveEntries[idx+1:]...)
+						inAddAndRemove = true
+						break
 					}
+				}
+
+				if inAddAndRemove {
+					descriptionOnlyRemovals = append(descriptionOnlyRemovals, removeEntry)
+				} else {
+					removals = append(removals, removeEntry)
 				}
 			}
 
-			if len(removalInput.RemoveEntries) > 0 {
-				_, err := conn.ModifyManagedPrefixList(removalInput)
+			if len(descriptionOnlyRemovals) > 0 {
+				_, err := conn.ModifyManagedPrefixList(&ec2.ModifyManagedPrefixListInput{
+					CurrentVersion: input.CurrentVersion,
+					PrefixListId:   aws.String(d.Id()),
+					RemoveEntries:  descriptionOnlyRemovals,
+				})
 
 				if err != nil {
 					return fmt.Errorf("error updating EC2 Managed Prefix List (%s): %w", d.Id(), err)
@@ -274,12 +284,14 @@ func resourceAwsEc2ManagedPrefixListUpdate(d *schema.ResourceData, meta interfac
 				}
 
 				input.CurrentVersion = managedPrefixList.Version
+			}
 
+			if len(removals) > 0 {
+				input.RemoveEntries = removals
+			} else {
 				// Prevent this error if RemoveEntries is list with no elements after removals:
 				//   InvalidRequest: The request received was invalid.
-				if len(input.RemoveEntries) == 0 {
-					input.RemoveEntries = nil
-				}
+				input.RemoveEntries = nil
 			}
 		}
 

--- a/aws/resource_aws_ec2_managed_prefix_list_test.go
+++ b/aws/resource_aws_ec2_managed_prefix_list_test.go
@@ -170,9 +170,13 @@ func TestAccAwsEc2ManagedPrefixList_Entry_Description(t *testing.T) {
 				ResourceName: resourceName,
 				Check: resource.ComposeAggregateTestCheckFunc(
 					testAccAwsEc2ManagedPrefixListExists(resourceName),
-					resource.TestCheckResourceAttr(resourceName, "entry.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "entry.#", "2"),
 					resource.TestCheckTypeSetElemNestedAttrs(resourceName, "entry.*", map[string]string{
 						"cidr":        "1.0.0.0/8",
+						"description": "description1",
+					}),
+					resource.TestCheckTypeSetElemNestedAttrs(resourceName, "entry.*", map[string]string{
+						"cidr":        "2.0.0.0/8",
 						"description": "description1",
 					}),
 					resource.TestCheckResourceAttr(resourceName, "version", "1"),
@@ -188,9 +192,13 @@ func TestAccAwsEc2ManagedPrefixList_Entry_Description(t *testing.T) {
 				ResourceName: resourceName,
 				Check: resource.ComposeAggregateTestCheckFunc(
 					testAccAwsEc2ManagedPrefixListExists(resourceName),
-					resource.TestCheckResourceAttr(resourceName, "entry.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "entry.#", "2"),
 					resource.TestCheckTypeSetElemNestedAttrs(resourceName, "entry.*", map[string]string{
 						"cidr":        "1.0.0.0/8",
+						"description": "description2",
+					}),
+					resource.TestCheckTypeSetElemNestedAttrs(resourceName, "entry.*", map[string]string{
+						"cidr":        "2.0.0.0/8",
 						"description": "description2",
 					}),
 					resource.TestCheckResourceAttr(resourceName, "version", "3"), // description-only updates require two operations
@@ -414,6 +422,11 @@ resource "aws_ec2_managed_prefix_list" "test" {
 
   entry {
     cidr        = "1.0.0.0/8"
+    description = %[2]q
+  }
+
+  entry {
+    cidr        = "2.0.0.0/8"
     description = %[2]q
   }
 }

